content/en/blog/_posts/2019-10-20-csi-ephemeral-inline-volumes.md

```diff
@@ -0,0 +1,75 @@
+---
+title: CSI Ephemeral Inline Volumes
+date: 2019-10-20
+---
+
+**Author:** Patrick Ohly (Intel)
+
+Typically, volumes provided by an external storage driver in
+Kubernetes are *persistent*, with a lifecycle that is completely
+independent of pods or (as a special case) loosely coupled to the
+first pod which uses a volume ([late binding
+mode](https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode)).
+The mechanism for requesting and defining such volumes in Kubernetes
+are [Persistent Volume Claim (PVC) and Persistent Volume
+(PV)](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
+objects. Originally, volumes that are backed by a Container Storage Interface
+(CSI) driver could only be used via this PVC/PV mechanism.
+
+But there are also use cases for data volumes whose content and
+lifecycle is tied to a pod. For example, a driver might populate a
+volume with dynamically created secrets that are specific to the
+application running in the pod. Such volumes need to be created
+together with a pod and can be deleted as part of pod termination
+(*ephemeral*). They get defined as part of the pod spec (*inline*).
+
+Since Kubernetes 1.15, CSI drivers can also be used for such
+*ephemeral inline* volumes. The [CSIInlineVolume feature
+gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
+had to be set to enable it in 1.15 because support was still in alpha
+state. In 1.16, the feature reached beta state, which typically means
+that it is enabled in clusters by default.
+
+CSI drivers have to be adapted to support this because although an
+existing CSI gRPC call (`NodePublishVolume`) is used, the way how it
+is used is different and not covered by the CSI spec: for ephemeral
+volumes, only that call is invoked by `kubelet` and all other calls
```

Comment:
Minor nit. NodeUnpublishVolume is also called.

Reply:
Indeed, will fix that.